### ISSUE TITLE
Add publisher/observer by primary key even if object doesn't exist in Realm yet

### ISSUE THREAD

**Issue by Uki19** [realm/realm-swift#7834]
### Problem

Common pattern I'm facing is setting up observer for single object then fetching and saving it from API. There are few difficulties I have when I use `valuePublisher(object)` / `object.observe { }`:
1. changes to relationship properties wont trigger observation block, it's mentioned in [this comment](https://github.com/realm/realm-swift/issues/5878#issuecomment-414421577) but afaik there is no option to opt-in to deep changes notifications yet

2. If object doesn't exist in realm yet, observation won't be set up since it's nil, causing few additional steps each time:
      - check if object exists and add observer
      - fetch from api
      - check if observation was added already, if not, add it to newly added object
      
3. Current existing workaround is using collection observer which requires using `Results`, `.filter` and `.first`:
```
extension Results where Element: Object {
    func objectPublisher<KeyType>(for primaryKey: KeyType) -> AnyPublisher<Element, Error> {
        var objects = self
        if let shema = Element.sharedSchema(),
           let primaryKeyProperty = shema.primaryKeyProperty {
            objects = filter("\(primaryKeyProperty.name) = %@", primaryKey)
        }

        return objects
            .collectionPublisher
            .compactMap(\.first)
            .eraseToAnyPublisher()
    }
} 
```

### Solution

The idea here is to have something similar to what code snippet above achieves, but without the use of filtered Results collection publisher. I'm wondering if this can be achieved with realm internal stuff and made as separate publisher/observer. Though it's preferred to work more like collectionPublisher since it's triggered for nested properties changes also.

Basically, it could look something like:
```
class Dog: Object {
@Persisted(primaryKey: true) var id: String
@Persisted var name = ""
}

objectNotificationToken = Dog.observe(withPrimaryKey: "123") { change in
switch change {
    case .created:
     // Object with this id was added to realm
    case .change(let object, let properties):
    // Properties changed, including nested properties
    case .error(let error):
    // Error
    case .deleted:
    // Object object was deleted
    }
}

// Publisher
Dog.publisher(withPrimaryKey: "123") 
```

### How important is this improvement for you?

I'd like to see it, but have a workaround

**Comment by ericjordanmossman** [comment]
I don't know if this^ addresses the "even if object doesn't exist in realm" part though.

**Comment by Uki19** [comment]
> > but afaik there is no option to opt-in to deep changes notifications yet
> 
> The comment you linked was correct at the time. There is now an option to [filter via key paths](https://github.com/realm/realm-swift/blob/master/CHANGELOG.md#enhancements-20), where you can see observations on nested properties, including change in relationship. See [here](https://github.com/realm/realm-swift/blob/master/RealmSwift/Object.swift#L261).
> 
> Let me know if I'm misunderstanding and this doesn't do what you need.

Yeah filtering by keypath triggers nested properties changes, but then if we wanted to trigger notification on each property and nested property change, we'd have huge array of keypaths we want to observe. I think it would be useful if there's an `observe` call on Object which works similarly to Results, as mentioned in linked comment: 
```
 Results notifications do notify when it is a linked object which changed (up to 4 levels deep).
```



> I don't know if this^ addresses the "even if object doesn't exist in realm" part though.

Yup, it doesn't. That part is kinda unrelated to nested change notifications, it's more about having `.created/.added` notification for object with primary key we want to observe
